File: controllers/xyuan/pass.go
package xyuan

import (
	"fmt"
	"github.com/lonySp/go-gin-shop-admin/models"
	"net/http"
	"regexp"
	"strings"

	"github.com/gin-contrib/sessions"
	"github.com/gin-gonic/gin"
)

type PassController struct {
	BaseController
}

// 获取验证码
func (con PassController) Captcha(c *gin.Context) {
	id, b64s, _, err := models.MackCaptcha(50, 120, 4)

	if err != nil {
		fmt.Println(err)
	}
	c.JSON(http.StatusOK, gin.H{
		"captchaId":    id,
		"captchaImage": b64s,
	})
}

func (con PassController) Login(c *gin.Context) {
	//生成随机数
	prevPage := c.Request.Referer()

	c.HTML(http.StatusOK, "xyuan/pass/login.html", gin.H{
		"prevPage": prevPage,
	})
}
func (con PassController) RegisterStep1(c *gin.Context) {
	c.HTML(http.StatusOK, "xyuan/pass/register_step1.html", gin.H{})
}
func (con PassController) RegisterStep2(c *gin.Context) {
	sign := c.Query("sign")
	verifyCode := c.Query("verifyCode")

	//1、验证图形验证码是否正确

	session := sessions.Default(c)
	sessionVerifyCode := session.Get("verifyCode")
	sessionVerifyCodeStr, ok := sessionVerifyCode.(string)
	if !ok || verifyCode != sessionVerifyCodeStr {
		c.Redirect(302, "/pass/registerStep1")
	}

	//2、获取sign 判断sign是否合法

	userTemp := []models.UserTemp{}
	models.DB.Where("sign=?", sign).Find(&userTemp)
	if len(userTemp) > 0 {
		c.HTML(http.StatusOK, "xyuan/pass/register_step2.html", gin.H{
			"phone":      userTemp[0].Phone,
			"verifyCode": verifyCode,
			"sign":       sign,
		})
	} else {
		c.Redirect(302, "/pass/registerStep1")
	}

}
func (con PassController) RegisterStep3(c *gin.Context) {

	sign := c.Query("sign")
	smsCode := c.Query("smsCode")

	//1、验证短信验证码是否正确
	session := sessions.Default(c)
	sessionSmsCode := session.Get("smsCode")
	sessionSmsCodeStr, ok := sessionSmsCode.(string)

	if !ok || smsCode != sessionSmsCodeStr {
		c.Redirect(302, "/pass/registerStep1")
	}

	//2、获取sign 判断sign是否合法

	userTemp := []models.UserTemp{}
	models.DB.Where("sign=?", sign).Find(&userTemp)
	if len(userTemp) > 0 {
		c.HTML(http.StatusOK, "xyuan/pass/register_step3.html", gin.H{
			"smsCode": smsCode,
			"sign":    sign,
		})
	} else {
		c.Redirect(302, "/pass/registerStep1")
	}

}

func (con PassController) DoRegister(c *gin.Context) {

	//1、获取表单传过来的数据

	sign := c.PostForm("sign")
	smsCode := c.PostForm("smsCode")
	password := c.PostForm("password")
	rpassword := c.PostForm("rpassword")

	//2、验证smsCode是否合法
	session := sessions.Default(c)
	sessionSmsCode := session.Get("smsCode")
	sessionSmsCodeStr, ok := sessionSmsCode.(string)

	if !ok || smsCode != sessionSmsCodeStr {
		c.Redirect(302, "/")
	}
	//3、验证密码是否合法
	if len(password) < 6 || password != rpassword {
		c.Redirect(302, "/")
	}
	//4、验证签名是否合法
	userTemp := []models.UserTemp{}
	models.DB.Where("sign=?", sign).Find(&userTemp)
	if len(userTemp) > 0 {
		//4、完成注册
		user := models.User{
			Phone:    userTemp[0].Phone,
			Password: models.Md5(password), //密码要加密
			LastIp:   userTemp[0].Ip,
			AddTime:  int(models.GetUnix()),
			Status:   1,
		}
		models.DB.Create(&user)

		//5、执行登录
		models.Cookie.Set(c, "userinfo", user)

		c.Redirect(302, "/")

	} else {
		c.Redirect(302, "/")
	}

}

func (con PassController) SendCode(c *gin.Context) {

	phone := c.Query("phone")
	verifyCode := c.Query("verifyCode")
	captchaId := c.Query("captchaId")

	if captchaId == "resend" {
		// 1、注册第二个页面发送验证码的时候需要验证图形验证码
		sessionDefault := sessions.Default(c)
		sessionVerifyCode := sessionDefault.Get("verifyCode")
		sessionVerifyCodeStr, ok := sessionVerifyCode.(string)
		if !ok || verifyCode != sessionVerifyCodeStr {
			c.JSON(http.StatusOK, gin.H{
				"success": false,
				"message": "非法请求",
			})
			return
		}
	} else {
		// 1、验证图形验证码是否正确 保存图形验证码
		if flag := models.VerifyCaptcha(captchaId, verifyCode); !flag {
			c.JSON(http.StatusOK, gin.H{
				"success": false,
				"message": "验证码输入错误，请重试",
			})
			return
		}
		//保存图形验证码
		sessionDefault := sessions.Default(c)
		sessionDefault.Set("verifyCode", verifyCode)
		sessionDefault.Save()
	}
	/*
		2、判断手机格式是否合法
				pattern := `^[\d]{11}$`
				reg := regexp.MustCompile(pattern)
				reg.MatchString(phone)
	*/
	pattern := `^[\d]{11}$`
	reg := regexp.MustCompile(pattern)
	if !reg.MatchString(phone) {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "手机号格式不合法",
		})
		return
	}

	//3、验证手机号是否注册过
	userList := []models.User{}
	models.DB.Where("phone = ?", phone).Find(&userList)
	if len(userList) > 0 {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "手机号已经注册，请直接登录",
		})
		return
	}
	//4、判断当前ip地址今天发送短信的次数

	ip := c.ClientIP()
	currentDay := models.GetDay() //20211211
	var sendCount int64
	models.DB.Table("user_temp").Where("ip=? AND add_day=?", ip, currentDay).Count(&sendCount)
	if sendCount > 4 {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "此ip今天发送短信的次数已经达到上限，请明天再试",
		})
		return
	}
	//5、验证当前手机号今天发送的次数是否合法
	userTemp := []models.UserTemp{}
	smsCode := models.GetRandomNum()
	sign := models.Md5(phone + currentDay) //签名：主要用于页面跳转传值
	models.DB.Where("phone = ? AND add_day=?", phone, currentDay).Find(&userTemp)
	if len(userTemp) > 0 {
		if userTemp[0].SendCount > 2 {
			c.JSON(http.StatusOK, gin.H{
				"success": false,
				"message": "此手机号今天发送短信的次数已经达到上限，请明天再试",
			})
			return
		} else {
			//1、生成短信验证码  发送验证码  调用前面课程的接口
			fmt.Println("----------自己集成发送短信的接口--------")
			fmt.Println(smsCode)
			//2、服务器保存验证码
			session := sessions.Default(c)
			session.Set("smsCode", smsCode)
			session.Save()

			//3、更新发送短信的次数
			oneUserTemp := models.UserTemp{}
			models.DB.Where("id=?", userTemp[0].Id).Find(&oneUserTemp)
			oneUserTemp.SendCount = oneUserTemp.SendCount + 1
			oneUserTemp.AddTime = int(models.GetUnix())

			models.DB.Save(&oneUserTemp)

			c.JSON(http.StatusOK, gin.H{
				"success": true,
				"message": "发送短信成功",
				"sign":    sign,
			})
			return
		}

	} else {
		//1、生成短信验证码  发送验证码  调用前面课程的接口
		fmt.Println("----------自己集成发送短信的接口--------")
		fmt.Println(smsCode)
		//2、服务器保存验证码
		session := sessions.Default(c)
		session.Set("smsCode", smsCode)
		session.Save()

		//3、记录发送短信的次数

		oneUserTemp := models.UserTemp{
			Ip:        ip,
			Phone:     phone,
			SendCount: 1,
			AddDay:    currentDay,
			AddTime:   int(models.GetUnix()),
			Sign:      sign,
		}
		models.DB.Create(&oneUserTemp)

		c.JSON(http.StatusOK, gin.H{
			"success": true,
			"message": "发送短信成功",
			"sign":    sign,
		})
		return

	}
}

// 验证验证码
func (con PassController) ValidateSmsCode(c *gin.Context) {

	sign := c.Query("sign")
	smsCode := c.Query("smsCode")
	//1、验证数据是否合法

	userTemp := []models.UserTemp{}
	models.DB.Where("sign=?", sign).Find(&userTemp)
	if len(userTemp) == 0 {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "非法请求",
		})
		return
	}

	//2、验证短信验证码是否正确
	sessionDefault := sessions.Default(c)
	sessionSmsCode := sessionDefault.Get("smsCode")
	sessionSmsCodeStr, ok := sessionSmsCode.(string)
	if !ok || smsCode != sessionSmsCodeStr {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "短信验证码输入错误",
		})
		return
	}
	//3、判断验证码有没有过期   15分

	nowTime := models.GetUnix()
	if (nowTime-int64(userTemp[0].AddTime))/1000/60 > 15 {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "短信验证码已过期",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "验证码输入正确",
	})
}

func (con PassController) DoLogin(c *gin.Context) {

	phone := strings.Trim(c.PostForm("phone"), " ")
	password := c.PostForm("password")
	captchaId := c.PostForm("captchaId")
	captchaVal := c.PostForm("captchaVal")

	//1、验证图形验证码是否合法
	if flag := models.VerifyCaptcha(captchaId, captchaVal); !flag {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "图形验证码不正确",
		})
		return
	}

	//2、验证用户名密码是否正确
	password = models.Md5(strings.Trim(password, " "))
	userList := []models.User{}
	models.DB.Where("phone = ? AND password = ?", phone, password).Find(&userList)
	if len(userList) > 0 {
		//执行登录
		models.Cookie.Set(c, "userinfo", userList[0])
		c.JSON(http.StatusOK, gin.H{
			"success": true,
			"message": "用户登录成功",
		})

	} else {
		c.JSON(http.StatusOK, gin.H{
			"success": false,
			"message": "用户名或者密码错误",
		})
		return
	}

}
func (con PassController) LoginOut(c *gin.Context) {
	//删除cookie里面的userinfo执行跳转
	models.Cookie.Remove(c, "userinfo")
	prevPage := c.Request.Referer()
	if len(prevPage) > 0 {
		c.Redirect(302, prevPage)
	} else {
		c.Redirect(302, "/")
	}
}

File: static/xyuan/js/pass.js
(function($){
    $(function(){
        loginApp.init();
    })
    var loginApp={
        init:function(){
            this.getCaptcha();
            this.captchaImgChage();
            this.initRegisterStep1();
			this.initRegisterStep2();
			this.initRegisterStep3();
			this.initDoLogin();
        },
        getCaptcha:function(){
            $.get("/pass/captcha?t="+Math.random(),function(response){              
                $("#captchaId").val(response.captchaId)
                $("#captchaImg").attr("src",response.captchaImage)
            })
        },
        captchaImgChage:function(){
            var _that=this;
            $("#captchaImg").click(function(){
                _that.getCaptcha()
            })
        },initRegisterStep1:function(){
            var _that=this;
            //发送验证码
			$("#registerButton").click(function () {
				//验证验证码是否正确
				var phone = $('#phone').val();
				var verifyCode = $('#verifyCode').val();
				var captchaId = $("#captchaId").val();
				$(".error").html("")

				var reg = /^[\d]{11}$/;
				if (!reg.test(phone)) {
					$(".error").html("Error：手机号输入错误");
					return false;
				}
				if (verifyCode.length < 2) {
					$(".error").html("Error：图形验证码长度不合法")
					return false;
				}						
				
				$.get("/pass/sendCode",{"phone":phone,"verifyCode":verifyCode,"captchaId":captchaId},function(response){
					console.log(response)
					if (response.success == true) {						
						//跳转到下页面
						location.href="/pass/registerStep2?sign="+response.sign+"&verifyCode="+verifyCode;				
					} else {
						//改变验证码											
						$(".error").html("Error：" + response.message + ",请重新输入!")
						//改变验证码
                        _that.getCaptcha()
												
					}
				})			

			})
        },initRegisterStep2:function(){
			$(function () {
				var timer = 10;
				function Countdown() {
					if (timer >= 1) {
						timer -= 1;
						$("#sendCode").attr('disabled', true);
						$("#sendCode").html('重新发送(' + timer + ')');
						setTimeout(function () {
							Countdown();
						}, 1000);
					} else {
						$("#sendCode").attr('disabled', false)
						$("#sendCode").html('重新发送');
					}
				}
				Countdown();

				$("#sendCode").click(function () {
					timer = 10;
					Countdown();
					var phone = $("#phone").val()
					var verifyCode = $("#verifyCode").val()
					var captchaId = "resend"

					//重新请求接口发送短信
					$.get("/pass/sendCode", { "phone": phone, "verifyCode": verifyCode, "captchaId": captchaId }, function (response) {
						console.log(response)
					})
				})
			})

			//验证验证码		
			$(function () {

				$("#nextStep").click(function (e) {
					$(".error").html()
					var sign = $('#sign').val();
					var smsCode = $('#smsCode').val();

					$.get('/pass/validateSmsCode', { sign: sign, smsCode: smsCode }, function (response) {
						console.log(response)
						if (response.success == true) {
							location.href = "/pass/registerStep3?sign=" + sign + "&smsCode=" + smsCode
						} else {
							$(".error").html("Error：" + response.message)
						}
					})

				})
				
				$("#returnButton").click(function(){
						location.href="/pass/registerStep1"
				})

			})
		},initRegisterStep3:function(){
			$(function(){
				$("#form").submit(function(){
					$(".error").html("")
					var password=$('#password').val();
					var rpassword=$('#rpassword').val();

					if(password.length<6){
						$(".error").html("Error：密码的长度不能小于6位")						
						return false;
					}					
					if(password!=rpassword){							
						$(".error").html("Error：密码和确认密码不一致")
						return false;
					}
					return true;

				})
			})
		},initDoLogin:function(){
			var _that=this;
			$("#doLogin").click(function(e){
				$(".error").html("")
				var phone=$('#phone').val();
				var password= $('#password').val();			
				var captchaId = $('#captchaId').val();
				var captchaVal = $("#captchaVal").val();
				var prevPage = $("#prevPage").val();
				var reg =/^[\d]{11}$/;
				if(!reg.test(phone)){
					$(".error").html('Error:手机号输入错误');					
					return false;
				}
				if(password.length<6){
					$(".error").html('Error:密码长度不合法');						
					return false;
				}

				if(captchaVal.length<4){						
					$(".error").html('Error:验证码长度不合法');	
					return false;
				}
				//ajax请求	 														
				$.post('/pass/doLogin',{phone:phone,password:password,captchaVal:captchaVal,captchaId:captchaId},function(response){							
					console.log(response);
					if(response.success==true){						
						if(prevPage == ""){
							location.href="/";
						}else{
							location.href=prevPage;
						}
					}else{															
						$(".error").html("Error：" + response.message + ",请重新输入!")
						//改变验证码	
						_that.getCaptcha();				
					}					
				})
			})
		}
    }
})($)

